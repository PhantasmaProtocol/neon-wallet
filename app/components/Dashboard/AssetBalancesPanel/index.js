// @flow
import { compose } from 'recompose'
import { values, omit } from 'lodash'
import { withActions } from 'spunky'

import AssetBalancesPanel from './AssetBalancesPanel'
import balancesActions from '../../../actions/balancesActions'
import withBalancesData from '../../../hocs/withBalancesData'
import withCurrencyData from '../../../hocs/withCurrencyData'
import withPricesData from '../../../hocs/withPricesData'
import withLoadingProp from '../../../hocs/withLoadingProp'
import withSuccessNotification from '../../../hocs/withSuccessNotification'
import withFailureNotification from '../../../hocs/withFailureNotification'

const mapBalanceDataToProps = (balances) => ({
  NEO: balances.NEO,
  GAS: balances.GAS,
  tokenBalances: values(omit(balances, 'NEO', 'GAS'))
})

const mapPricesDataToProps = ({ NEO, GAS }) => ({
  neoPrice: NEO,
  gasPrice: GAS
})

const mapBalancesActionsToProps = (actions, props) => ({
  refresh: () => actions.call({ net: props.net, address: props.address, tokens: props.tokens })
})

export default compose(
  withCurrencyData('currencyCode'),

  // Fetch prices data based based upon the selected currency.  Reload data with the currency changes.
  withPricesData(mapPricesDataToProps),

  // Fetch balances data based based upon the selected network.  Reload data with the network changes.
  withBalancesData(mapBalanceDataToProps),

  // Expose data & functionality needed for `refresh` action.
  withActions(balancesActions, mapBalancesActionsToProps),
  withLoadingProp(balancesActions),
  withSuccessNotification(balancesActions, 'Received latest blockchain information.'),
  withFailureNotification(balancesActions, 'Failed to retrieve blockchain information.')
)(AssetBalancesPanel)
