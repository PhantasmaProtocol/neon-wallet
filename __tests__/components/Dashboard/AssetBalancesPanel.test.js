import React from 'react'
import nock from 'nock'
import { merge } from 'lodash'
import { mount, shallow } from 'enzyme'
import { progressValues } from 'spunky'

import { createStore, provideStore, provideState } from '../../testHelpers'
import AssetBalancesPanel from '../../../app/components/Dashboard/AssetBalancesPanel'
import { DEFAULT_CURRENCY_CODE, MAIN_NETWORK_ID } from '../../../app/core/constants'

const { LOADED } = progressValues

const initialState = {
  spunky: {
    NETWORK: {
      batch: false,
      progress: LOADED,
      data: MAIN_NETWORK_ID,
      loadedCount: 1
    },
    AUTH: {
      batch: false,
      progress: LOADED,
      data: {
        address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
      },
      loadedCount: 1
    },
    SETTINGS: {
      batch: false,
      progress: LOADED,
      data: {
        currency: DEFAULT_CURRENCY_CODE,
        tokens: []
      },
      loadedCount: 1
    },
    PRICES: {
      batch: false,
      progress: LOADED,
      data: {
        NEO: 25.48,
        GAS: 18.1
      },
      loadedCount: 1
    },
    BALANCES: {
      batch: false,
      progress: LOADED,
      data: {
        NEO: '100001',
        GAS: '1000.0001601'
      },
      loadedCount: 1
    },
    CLAIMS: {
      batch: false,
      progress: LOADED,
      data: {
        total: '0.5'
      },
      loadedCount: 1
    }
  },
  claim: {
    claimRequest: false,
    disableClaimButton: false
  }
}

describe('AssetBalancesPanel', () => {
  beforeEach(() => {
    const response = [{ symbol: 'NEO', price_usd: 24.5 }, { symbol: 'GAS', price_usd: 18.2 }]
    nock('https://api.coinmarketcap.com')
      .get('/v1/ticker/')
      .query({ limit: 0, convert: 'USD' })
      .reply(200, response, { 'Access-Control-Allow-Origin': '*' })
  })

  test('renders without crashing', () => {
    const store = createStore(initialState)
    const wrapper = shallow(<AssetBalancesPanel store={store} />)
    expect(wrapper).toMatchSnapshot()
  })

  test.only('correctly renders data from state', () => {
    const wrapper = mount(provideState(<AssetBalancesPanel />, initialState))

    console.log('====================== HTML!!! ======================')
    console.log(wrapper.html())
    console.log('====================== HTML!!! ======================')

    expect(wrapper.find('#neoWalletValue').text()).toEqual(`$2,548,025.48`)
    expect(wrapper.find('#gasWalletValue').text()).toEqual(`$18,100.00`)
    expect(wrapper.find('#walletTotal').text()).toEqual(`$2,566,125.48`)
    expect(wrapper.find('#amountNeo').text()).toEqual('100,001')
    expect(wrapper.find('#amountGas').text()).toEqual('1,000.0002')
  })

  test('account data refreshes when refresh button is clicked', () => {
    const store = createStore(initialState)
    const wrapper = mount(provideStore(<AssetBalancesPanel />, store))

    wrapper.find('#refresh').hostNodes().simulate('click')

    expect(store.getActions()).toContainEqual(expect.objectContaining({
      type: 'BALANCES/ACTION/CALL',
      meta: expect.objectContaining({ id: 'BALANCES' })
    }))
  })

  test('correctly renders data from state with non-default currency', () => {
    const state = merge(initialState, {
      spunky: {
        SETTINGS: { data: { currency: 'eur' } },
        PRICES: { data: { NEO: 1.11, GAS: 0.55 } }
      }
    })
    const wrapper = mount(provideState(<AssetBalancesPanel />, state))

    expect(wrapper.find('#neoWalletValue').text()).toEqual(`€111,001.11`)
    expect(wrapper.find('#gasWalletValue').text()).toEqual(`€550.00`)
    expect(wrapper.find('#walletTotal').text()).toEqual(`€111,551.11`)
  })
})
