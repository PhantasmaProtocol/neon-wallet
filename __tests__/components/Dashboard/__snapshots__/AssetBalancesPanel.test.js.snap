// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetBalancesPanel renders without crashing 1`] = `
<withProgress(ConditionalCallComponent)
  __progress__="LOADED"
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    t {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
